apps/shipments/serializers.py

```diff
@@ -125,11 +123,24 @@ def create(self, validated_data):
                         data['owner_id'] = validated_data['owner_id']
                     extra_args[location_field], _ = Location.objects.get_or_create(**data)
 
-            if 'device_id' in validated_data:
-                extra_args['device'] = Device.get_or_create_with_permission(auth, validated_data.pop('device_id'))
+            if 'device_id' in self.context:
+                extra_args['device'] = self.context['device']
 
             return Shipment.objects.create(**validated_data, **extra_args)
 
+    def validate_device_id(self, device_id):
+        auth = self.context['auth']
+
+        device = Device.get_or_create_with_permission(auth, device_id)
+        if hasattr(device, 'shipment'):
+            if not device.shipment.delivery_act:
+                raise serializers.ValidationError('Device is already assigned to a Shipment in progress')
+            else:
+                Shipment.objects.filter(device_id=device.id).update(device=None)
```

Comment:
It shouldn't be?
`Shipment.objects.filter(device__id=device.id).update(device=None)`

Reply:
I think either would work, really. The OneToOne/ForeignKey model field actually adds a device_id column to the shipment table, so I think it's fine to reference that directly:
https://docs.djangoproject.com/en/2.1/ref/models/fields/#database-representation